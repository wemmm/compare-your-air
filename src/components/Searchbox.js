const Searchbox = ({ props }) => {
  return (
    <input
      className="searchbox"
      type="text"
      name="myCountry"
      placeholder="Enter city name..."
    ></input>
  );
};

export default Searchbox;
