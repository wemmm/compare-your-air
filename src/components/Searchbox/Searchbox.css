.searchbox {
  border-color: #cccccc;
  border-style: solid;
  border-radius: 8px;
  border-width: 2px;
  box-sizing: border-box;
  background: url("../../svg/search.svg") no-repeat scroll 7px 7px, white;
  padding: 9px;
  padding-left: 40px;
  font-size: 15px;
  width: 100%;
  max-width: 400px;
  position: relative;
  z-index: 1;
}

.searchbox::placeholder {
  color: black;
}

.searchbox--suggestion-list {
  background-color: white;
  list-style: none;
  width: 100%;
  max-width: 400px;
  max-height: 210px;
  margin-left: auto;
  margin-right: auto;
  padding-left: 0;
  text-align: left;
  position: relative;
  bottom: 22px;
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
  overflow: scroll;
}

.searchbox--suggestion {
  padding: 10px 15px;
  cursor: pointer;
  font-family: "Open Sans", sans-serif;
}

.searchbox--suggestion:first-of-type {
  padding-top: 15px;
}

.searchbox--suggestion:hover {
  background-color: #f2f2f2;
  overflow: hidden;
}
