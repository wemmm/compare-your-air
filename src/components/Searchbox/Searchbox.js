import "./Searchbox.css";
import { useState } from "react";

const Searchbox = ({ placeholder, options, onSuggestionClick }) => {
  const [suggestions, setSuggestions] = useState([]);

  const onChange = (e) => {
    const query = e.target.value;

    if (query) {
      setSuggestions(
        options.filter(
          (option) =>
            option.toLowerCase().includes(query.toLowerCase()) &&
            option[0].toLowerCase() === query[0].toLowerCase()
        )
      );
    } else {
      setSuggestions([]);
    }
  };

  return (
    <form>
      <input
        className="searchbox"
        type="text"
        placeholder={placeholder}
        onChange={onChange}
      />

      {suggestions.length > 0 ? (
        <ul className="searchbox--suggestion-list">
          {suggestions.map((suggestion, i) => (
            <li
              key={i}
              className="searchbox--suggestion"
              onClick={() => onSuggestionClick(suggestion)}
            >
              {suggestion}
            </li>
          ))}
        </ul>
      ) : (
        false
      )}
    </form>
  );
};

export default Searchbox;
