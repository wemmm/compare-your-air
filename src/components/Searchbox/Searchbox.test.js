import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import Searchbox from "./Searchbox";

const options = ["Apple", "Banana", "Cherry"];

test("renders", () => {
  const component = renderer.create(
    <Searchbox placeholder="Placeholder" options={options} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders an autocompleted suggestion of options based on user input", () => {
  render(<Searchbox placeholder="Placeholder" options={options} />);

  const input = screen.getByPlaceholderText("Placeholder");
  fireEvent.change(input, { target: { value: "App" } });

  expect(input.value).toBe("App");
  expect(screen.getByText("Apple")).toBeInTheDocument();
  expect(screen.queryByText("Banana")).not.toBeInTheDocument();
  expect(screen.queryByText("Cherry")).not.toBeInTheDocument();
  expect(screen.getAllByRole("listitem").length).toBe(1);
});

test("is not case sensitive", () => {
  render(<Searchbox placeholder="Placeholder" options={options} />);

  const input = screen.getByPlaceholderText("Placeholder");
  fireEvent.change(input, { target: { value: "app" } });

  expect(input.value).toBe("app");
  expect(screen.getByText("Apple")).toBeInTheDocument();
  expect(screen.queryByText("Banana")).not.toBeInTheDocument();
  expect(screen.queryByText("Cherry")).not.toBeInTheDocument();
  expect(screen.getAllByRole("listitem").length).toBe(1);
});

test("can have multiple matches", () => {
  const options = ["Apple", "Apricot", "Cherry"];

  render(<Searchbox placeholder="Placeholder" options={options} />);

  const input = screen.getByPlaceholderText("Placeholder");
  fireEvent.change(input, { target: { value: "Ap" } });

  expect(input.value).toBe("Ap");
  expect(screen.getByText("Apple")).toBeInTheDocument();
  expect(screen.getByText("Apricot")).toBeInTheDocument();
  expect(screen.queryByText("Cherry")).not.toBeInTheDocument();
  expect(screen.getAllByRole("listitem").length).toBe(2);
});

test("can have no matches", () => {
  render(<Searchbox placeholder="Placeholder" options={options} />);

  const input = screen.getByPlaceholderText("Placeholder");
  fireEvent.change(input, { target: { value: "Orange" } });

  expect(input.value).toBe("Orange");
  expect(screen.queryByText("Apple")).not.toBeInTheDocument();
  expect(screen.queryByText("Banana")).not.toBeInTheDocument();
  expect(screen.queryByText("Cherry")).not.toBeInTheDocument();
  expect(screen.queryAllByRole("listitem").length).toBe(0);
});

test("removes suggestion list if input is empty", () => {
  const options = ["Apple", "Apricot", "Cherry"];

  render(<Searchbox placeholder="Placeholder" options={options} />);

  const input = screen.getByPlaceholderText("Placeholder");
  fireEvent.change(input, { target: { value: "Ap" } });

  expect(screen.getAllByRole("listitem").length).toBe(2);

  fireEvent.change(input, { target: { value: "" } });

  expect(screen.queryAllByRole("listitem").length).toBe(0);
});
