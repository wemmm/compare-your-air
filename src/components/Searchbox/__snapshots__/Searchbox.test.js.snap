// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<form>
  <input
    className="searchbox"
    onChange={[Function]}
    placeholder="Placeholder"
    type="text"
  />
</form>
`;
